### ISSUE TITLE
Dart lang improvements

### ISSUE THREAD

**Issue by dgrove** [dart-lang/language#582]
@pouloghost commented on [Tue Sep 17 2019](https://github.com/dart-lang/sdk/issues/38446)

I'm new to flutter/dart from android. Dart seems to be more java than kotlin, which makes dart quite outdated. Some great kotlin language features are missing, and will you consider adding them into new versions of dart?

- immutable at language level ( data class in kotlin)
- property delegation ( by KProperty in kotlin)
- inheritance limitation ( default not inheritable)
- more declarative generic (where clause in C#)

**Comment by natebosch** [comment]
> immutable at language level 

Please follow https://github.com/dart-lang/language/issues/314 and https://github.com/dart-lang/language/issues/125, and maybe https://github.com/dart-lang/language/issues/546

> inheritance limitation

We do have `@sealed`. It's analyzer only and not at the language level. See https://pub.dev/documentation/meta/latest/meta/sealed-constant.html

I'm not sure if we have tracking for a request to make this language level - @leafpetersen do you know of any?

**Comment by leafpetersen** [comment]
> I'm not sure if we have tracking for a request to make this language level - @leafpetersen do you know of any?

We don't seem to have one.

**Comment by natebosch** [comment]
#349 and #704 cover sealing.

> * property delegation ( by KProperty in kotlin)

IIUC this is runtime reflection and not something we plan to add.

> * more declarative generic (where clause in C#)

I'm guessing this is a request for LINQ? I think we'd probably want a more detail description of the use case for something this big.

This issue does not seem actionable outside of the linked duplicates. I'll close for now.

**Comment by mateusfccp** [comment]
@natebosch The "more declarative generic (where clause in C#)" seems to be referring to the [generic type constraint](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/where-generic-type-constraint) feature.

We already support some of this, like `A<T extends B>`, but not everything. One of the differences that I think is valuable is when constructor constraints, like:
```csharp
public class MyGenericClass<T> where T : IComparable<T>, new()
{
    // The following line is not possible without new() constraint:
    T item = new T();
}
```

There is already an issue for this, although closed: #30074.

**Comment by natebosch** [comment]
Thanks for the clarification!

There is also some discussion about being able to use some static interface from a generic type variable at https://github.com/dart-lang/language/issues/356
